File: src/main/java/me/lukiiy/utils/system/Vanish.java
package me.lukiiy.utils.system;

import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
import me.lukiiy.utils.cool.PlayerHelper;
import me.lukiiy.utils.cool.Presets;
import me.lukiiy.utils.main;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.server.ServerListPingEvent;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.UUID;

public class Vanish implements CommandExecutor, Listener {
    public static List<UUID> vanish = new ArrayList<>();

    @Override
    public boolean onCommand(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {
        if (!(commandSender instanceof Player) && strings.length == 0) {
            commandSender.sendMessage(main.argsErrorMsg);
            return true;
        }
        Player target = strings.length > 0 ? Bukkit.getPlayer(strings[0]) : (Player) commandSender;
        if (target == null) {
            commandSender.sendMessage(main.notFoundMsg);
            return true;
        }

        boolean invisible = vanish.contains(target.getUniqueId());
        Component state = invisible ? main.OFF : main.ON;

        Component message = Presets.Companion.msg("Vanish is now ").append(state);
        if (target != commandSender) {
            commandSender.sendMessage(message.append(Presets.Companion.why(" for ")).append(target.name().color(Presets.Companion.getACCENT_NEUTRAL())));
            message = message.append(Presets.Companion.why(" (by " + commandSender.getName() + ")"));
        }

        if (!invisible) {
            hide(target);
            vanish.add(target.getUniqueId());
        } else {
            show(target);
            vanish.remove(target.getUniqueId());
        }
        target.sendMessage(message);
        return true;
    }

    private void hide(Player p) {
        Bukkit.getOnlinePlayers().stream()
                .filter(online -> !p.equals(online))
                .forEach(online -> online.hidePlayer(main.plugin, p));
    }

    private void show(Player p) {
        Bukkit.getOnlinePlayers().stream()
                .filter(online -> !p.equals(online))
                .forEach(online -> online.showPlayer(main.plugin, p));
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void join(PlayerJoinEvent e) {
        if (vanish.isEmpty()) return;
        Player p = e.getPlayer();

        if (vanish.contains(p.getUniqueId())) {
            e.joinMessage(null);
            hide(p);
            p.sendMessage(Presets.Companion.msg("You are still vanished."));
        }

        Bukkit.getScheduler().runTaskLater(main.plugin, () -> {
            for (UUID vanished : vanish) {
                Player v = Bukkit.getPlayer(vanished);
                if (v != null) p.hidePlayer(main.plugin, v);
            }
        }, 2L);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void quit(PlayerQuitEvent e) {
        if (vanish.isEmpty()) return;
        Player p = e.getPlayer();
        if (vanish.contains(p.getUniqueId())) e.quitMessage(null);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void noListing(PaperServerListPingEvent e) {
        List<PaperServerListPingEvent.ListedPlayerInfo> samples = e.getListedPlayers();
        samples.removeIf(p -> vanish.contains(p.id()));
        e.setNumPlayers(samples.size());
    }
}

File: src/main/java/me/lukiiy/utils/cmd/Vanish.kt
package me.lukiiy.utils.cmd

import com.destroystokyo.paper.event.server.PaperServerListPingEvent
import com.mojang.brigadier.Command
import com.mojang.brigadier.tree.LiteralCommandNode
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.Commands
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import me.lukiiy.utils.Defaults
import me.lukiiy.utils.Lukitils
import me.lukiiy.utils.help.Utils.asPermission
import me.lukiiy.utils.help.Utils.getPlayerOrThrow
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.event.ClickEvent
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerAdvancementDoneEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.util.*

object Vanish : Listener {
    private val req: (CommandSender) -> Boolean = { it.hasPermission("vanish".asPermission()) }
    private val vanished = mutableSetOf<UUID>()

    fun getVanished(): Set<UUID> = vanished

    private val main = Commands.literal("vanish")
        .requires { req(it.sender) }
        .then(Commands.argument("player", ArgumentTypes.player())
            .executes {
                val sender = it.source.sender
                val target = it.getPlayerOrThrow("player")

                handle(sender, target)
                Command.SINGLE_SUCCESS
            })
        .executes {
            val sender = it.source.sender as? Player ?: throw Defaults.NON_PLAYER

            handle(sender, sender)
            Command.SINGLE_SUCCESS
        }

    fun toggle(target: Player): Boolean {
        val isVanished = target.uniqueId in vanished
        val players = Bukkit.getOnlinePlayers().minus(target)

        if (isVanished) {
            players.forEach { it.showPlayer(Lukitils.getInstance(), target) }
            vanished.remove(target.uniqueId)
        } else {
            players.forEach { it.hidePlayer(Lukitils.getInstance(), target) }
            vanished.add(target.uniqueId)
        }

        target.isSleepingIgnored = !isVanished
        return !isVanished
    }

    private fun handle(sender: CommandSender, target: Player) {
        val vanish = toggle(target)
        var message = Defaults.neutral(Component.text("Vanish is now ").append(if (vanish) Defaults.ON else Defaults.OFF))

        if (target != sender) {
            sender.sendMessage(message.append(Component.text(" for ").color(Defaults.GRAY)).append(target.name().color(Defaults.YELLOW)))
            message = message.append(Component.text(" (by ${sender.name})").color(Defaults.GRAY))
        }

        if (!Lukitils.getInstance().config.getBoolean("silentStats", true)) target.sendMessage(message)
    }

    private val list = Commands.literal("vanishlist")
        .requires { req(it.sender) }
        .executes {
            val sender = it.source.sender
            if (vanished.isEmpty()) throw Defaults.CmdException(Component.text("No vanished players found"))

            val players = vanished.mapNotNull { p -> Bukkit.getPlayer(p)?.name() }
            if (players.isEmpty()) throw Defaults.CmdException(Component.text("No online vanished players found"))

            sender.sendMessage(Defaults.neutral(Component.text("Vanished players:").appendNewline().append(Component.join(Defaults.LIST_LIKE, players.map { c -> c.color(Defaults.YELLOW) }))))
            Command.SINGLE_SUCCESS
        }

    fun register(): LiteralCommandNode<CommandSourceStack> = main.build()
    fun registerList(): LiteralCommandNode<CommandSourceStack> = list.build()

    // Listener
    @EventHandler(priority = EventPriority.LOWEST)
    fun join(e: PlayerJoinEvent) {
        if (vanished.isEmpty()) return
        val p = e.player

        if (vanished.contains(p.uniqueId)) {
            e.joinMessage(null)
            Bukkit.getOnlinePlayers().stream().filter { player: Player? -> player != p }.forEach { player: Player? -> player!!.hidePlayer(Lukitils.getInstance(), p) }
            p.sendMessage(Defaults.neutral(Component.text("You're still vanished!").clickEvent(ClickEvent.suggestCommand("/vanish"))))
        }

        Bukkit.getGlobalRegionScheduler().run(Lukitils.getInstance()) { _ ->
            vanished.stream().map { id -> Bukkit.getPlayer(id!!) }.filter { Objects.nonNull(it) }.forEach { player -> p.hidePlayer(Lukitils.getInstance(), player!!) }
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun quit(e: PlayerQuitEvent) {
        if (vanished.contains(e.player.uniqueId)) e.quitMessage(null)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun ping(e: PaperServerListPingEvent) {
        e.listedPlayers.removeIf { vanished.contains(it!!.id()) }
        e.numPlayers = e.listedPlayers.size
    }

    @EventHandler
    fun advancement(e: PlayerAdvancementDoneEvent) {
        if (vanished.contains(e.player.uniqueId)) e.message(null)
    }
}